ui/src/app/pages/data/data-set/dataset-page.tsx

```diff
@@ -519,60 +530,73 @@ const DataSetPage = fp.flow(withCurrentWorkspace(), withUrlParams())(
-      };
-      try {
-        const dataSetPreviewResp = await dataSetApi().previewQuery(namespace, id, request);
-        this.setState({
-          previewList: dataSetPreviewResp.domainValue,
-          selectedPreviewDomain: dataSetPreviewResp.domainValue[0].domain
-        });
-      } catch (ex) {
-        const exceptionResponse = await ex.json() as unknown as ErrorResponse;
-        let errorText: string;
-        switch (exceptionResponse.statusCode) {
-          case 400:
-            if (exceptionResponse.message ===
-              'Data Sets must include at least one cohort and concept.') {
-              errorText = exceptionResponse.message;
-            } else if (exceptionResponse.message ===
-              'Concept Sets must contain at least one concept') {
-              errorText = 'One or more of your concept sets has no concepts. ' +
-                'Please check your concept sets to ensure all concept sets have concepts.';
-            }
-            break;
-          case 404:
-            if (exceptionResponse.message.startsWith(
-              'Not Found: No Cohort definition matching cohortId')) {
-              errorText = 'Error with one or more cohorts in the data set. ' +
-                'Please submit a bug using the contact support button';
-            }
-            break;
-          case 504:
-            if (exceptionResponse.message ===
-              'Timeout while querying the CDR to pull preview information.') {
-              errorText = 'Query to load data from the All of Us Database timed out. ' +
-                'Please either try again or export data set to a notebook to try there';
-            }
-            break;
-          default:
-            errorText = 'An unexpected error has occurred. ' +
-              'Please submit a bug using the contact support button';
+      const domains = fp.uniq(this.state.selectedValues.map(domainValue => domainValue.domain));
+      this.setState({
+        previewList: domains.map(domain => {
+          return {
+            domain: domain,
+            isLoading: true,
+            values: []
+          };
+        }),
+        previewDataLoading: true,
+        selectedPreviewDomain: domains[0]
+      });
+      domains.forEach(async domain => {
+        const request = {
+          domain: domain,
+          conceptSetIds: this.state.selectedConceptSetIds,
+          includesAllParticipants: this.state.includesAllParticipants,
+          cohortIds: this.state.selectedCohortIds,
+          prePackagedConceptSet: this.getPrePackagedConceptSet(),
+          values: this.state.selectedValues.map(domainValue => domainValue.value)
+        };
+        try {
+          const dataSetPreviewResp = await dataSetApi().generateDataSetPreview(namespace, id, request);
+          const domainPreviewIndex = '[' + fp.findIndex(['domain', domain], this.state.previewList) + ']';
+          this.setState({previewList: fp.set(domainPreviewIndex + '.isLoading', false,
```

Comment:
I actually tried using that, the problem is it also has to pick out the right domain from the list, which is what the above code is doing. I can add a comment, and poke around a bit, but the finding the right object is the trickier part here, since previewList is a list. This could be an argument to move to a map.